tests/test_agent.py

```diff
@@ -3163,6 +3163,100 @@ def deferred_tool(x: int) -> int:  # pragma: no cover
             ]
         )
 
+    def test_exhaustive_strategy_calls_all_output_tools(self):
+        """Test that 'exhaustive' strategy executes all output tool functions."""
+        output_tools_called: list[str] = []
+
+        class FirstOutput(BaseModel):
+            value: str
+
+        class SecondOutput(BaseModel):
+            value: str
+
+        def process_first(output: FirstOutput) -> FirstOutput:
+            """Process first output."""
+            output_tools_called.append('first')
+            return output
+
+        def process_second(output: SecondOutput) -> SecondOutput:
+            """Process second output."""
+            output_tools_called.append('second')
+            return output
+
+        def return_model(_: list[ModelMessage], info: AgentInfo) -> ModelResponse:
+            assert info.output_tools is not None
+            return ModelResponse(
+                parts=[
+                    ToolCallPart('first_output', {'value': 'first'}),
+                    ToolCallPart('second_output', {'value': 'second'}),
```

Comment:
Can we also test here what happens if the second call is invalid? It should be consistent with exhaustive execution of non-output tool calls in terms of whether it causes us to go back to the model to make it try again, or whether we ignore the tool call failure because we already have valid final output

Reply:
https://github.com/pydantic/pydantic-ai/blob/5afc2d6fe67d61459ec80e738ea3e11aa1230c97/pydantic_ai_slim/pydantic_ai/_agent_graph.py#L1079-L1106
Judging by those lines, it seems retries are done to model in `exhaustive` strategy, if any tool (not output) is failed

Here's a small gift that confirms this behavior:
https://gist.github.com/Danipulok/0897bf27c1214adb7d4a401a684b0c39

Reply:
Ok great. Can you add a test here please for that? We have one for streaming, but I think we should have one here as well